message_ix/macro.py

```diff
@@ -1,61 +1,494 @@
+        self._growth()
+        self._rho()
+        self._gdp0()
+        self._k0()
+        self._total_cost()
+        self._price()
+        self._demand()
+        self._bconst()
+        self._aconst()
+
+    @lru_cache()
+    def _growth(self):
+        gdp = self.data['gdp_calibrate']
+        diff = gdp.groupby(level='node').diff()
+        years = sorted(gdp.index.get_level_values('year').unique())
+        dt = pd.Series(years, index=pd.Index(years, name='year')).diff()
+        growth = (diff / gdp + 1) ** (1 / dt) - 1
+        growth.name = 'value'
+        self.data['growth'] = growth.dropna()
+        return self.data['growth']
+
+    @lru_cache()
+    def _rho(self):
+        esub = self.data['esub']
+        self.data['rho'] = (esub - 1) / esub
+        return self.data['rho']
+
+    @lru_cache()
+    def _gdp0(self):
+        gdp = self.data['gdp_calibrate']
+        gdp0 = gdp.iloc[gdp.index.isin([self.init_year], level='year')]
+        self.data['gdp0'] = gdp0
+        return self.data['gdp0']
+
+    @lru_cache()
+    def _k0(self):
+        kgdp = self.data['kgdp']
+        gdp0 = self._gdp0()
+        self.data['k0'] = kgdp * gdp0
+        return self.data['k0']
+
+    @lru_cache()
+    def _total_cost(self):
+        # read from scenario
+        idx = ['node', 'year']
+        model_cost = self._clean_model_data(self.s.var('COST_NODAL_NET'))
+        model_cost.rename(columns={'lvl': 'value'}, inplace=True)
+        model_cost = model_cost[idx + ['value']]
+        # TODO: in the R code, this value is divided by 1000
+        # do we need to do that here?!!?
+        model_cost['value'] /= 1e3
+        # get data provided in init year from data
+        cost_ref = self.data['cost_ref'].reset_index()
+        cost_ref['year'] = self.init_year
+        # combine into one value
+        total_cost = pd.concat([cost_ref, model_cost]).set_index(idx)['value']
+        if total_cost.isnull().any():
+            raise RuntimeError('NaN values found in total_cost calculation')
+        self.data['total_cost'] = total_cost
+        return total_cost
+
+    @lru_cache()
+    def _price(self):
+        # read from scenario
+        idx = ['node', 'sector', 'year']
+        model_price = self._clean_model_data(
+            self.s.var('PRICE_COMMODITY', filters={'level': 'useful'})
+        )
+        if np.isclose(model_price['lvl'], 0).any():
+            # TODO: this needs a test..
+            msg = '0-price found in MESSAGE variable PRICE_COMMODITY'
+            raise RuntimeError(msg)
+        model_price.rename(columns={'lvl': 'value', 'commodity': 'sector'},
+                           inplace=True)
+        model_price = model_price[idx + ['value']]
+
+        # get data provided in init year from data
+        price_ref = self.data['price_ref'].reset_index()
+        price_ref['year'] = self.init_year
+        # combine into one value
+        price = pd.concat([price_ref, model_price]).set_index(idx)['value']
```

Comment:
A minor suggestion to turn off the warnings:
```
price = pd.concat([price_ref, model_price], sort=True).set_index(idx)['value']
```

Reply:
As of pandas 1.0.0, the warning is no longer raised; new users installing from scratch will not see this, so no need to explicitly give the argument to avoid a visible warning.